ConnectToExternalTargetsIT does not properly clean up

```
160026 < ebaron> aazores, does this test failure look familar? looks like an additional target showed up in one of the tests. https://github.com/cryostatio/cryostat/pull/500/checks?check_run_id=2749134320
160849 < aazores> ebaron: no, that one doesn't look familiar. Not sure why that would have happened
161831 < aazores> ebaron: weird, that itest passes for me on v1 locally still
161913 < aazores> ah, interesting though - the CI runner runs the itests in a different order from what I had locally
161952 < aazores> the itests are supposed to tear down their state after completion so that the next test has a clean slate to start from, so there might be a bug or a timing issue with that
162006 < ebaron> that's what I'm thinking as well
162049 < aazores> would make sense, the ConnectToExternalTargetsIT is the one that spins up other containers in the pod to connect to, so if that runs first and doesn't finish tearing down before the BasicCommandChannelIT starts then that 
                  would cause this failure
162138 < aazores> I guess the @AfterAll method in ConnectToExternalTargetsIT needs to check and wait for all the other test containers to actually be stopped/destroyed before returning
162240 < ebaron> there's a podman rm --force right before it that looks like it completed successfully, you'd think that would be enough
162408 < aazores> clearly I did think so at the time I wrote this haha
162417 < aazores> `podman rm --help` doesn't actually say this is synchronous though
162503 < aazores> (that podman rm --force is actually from the @AfterAll method, I think)
```

```
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running itest.GrafanaDashboardIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.18 s - in itest.GrafanaDashboardIT
[INFO] Running itest.ConnectToExternalTargetsIT
[podman, run, --quiet, --pod=cryostat-itests, --detach, --rm, quay.io/andrewazores/vertx-fib-demo:0.3.0]
d73e9f87c1e508563eca7dad129ae965d1bf72a3c5d69b5407d02a83f05212f6
[podman, container, inspect, --format="{{.State.Status}}", d73e9f87c1e508563eca7dad129ae965d1bf72a3c5d69b5407d02a83f05212f6]
"running"
[podman, rm, --force, d73e9f87c1e508563eca7dad129ae965d1bf72a3c5d69b5407d02a83f05212f6]
d73e9f87c1e508563eca7dad129ae965d1bf72a3c5d69b5407d02a83f05212f6
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 14.545 s - in itest.ConnectToExternalTargetsIT
[INFO] Running itest.BasicCommandChannelIT
Jun 04, 2021 7:53:45 PM io.vertx.core.impl.ContextImpl
SEVERE: Unhandled exception
java.lang.AssertionError: 
Expected: <[{"connectUrl":"service:jmx:rmi:///jndi/rmi://cryostat-itests:9091/jmxrmi","alias":"io.cryostat.Cryostat"}]>
     but: was <[{"connectUrl":"service:jmx:rmi:///jndi/rmi://cryostat-itests:9093/jmxrmi","alias":"es.andrewazor.demo.Main"},{"connectUrl":"service:jmx:rmi:///jndi/rmi://cryostat-itests:9091/jmxrmi","alias":"io.cryostat.Cryostat"}]>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
	at itest.BasicCommandChannelIT.lambda$shouldGetScanTargetsResponse$2(BasicCommandChannelIT.java:125)
	at io.vertx.ext.web.client.impl.HttpContext.handleDispatchResponse(HttpContext.java:313)
	at io.vertx.ext.web.client.impl.HttpContext.execute(HttpContext.java:300)
	at io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:275)
	at io.vertx.ext.web.client.impl.predicate.PredicateInterceptor.handle(PredicateInterceptor.java:70)
	at io.vertx.ext.web.client.impl.predicate.PredicateInterceptor.handle(PredicateInterceptor.java:32)
	at io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:272)
	at io.vertx.ext.web.client.impl.HttpContext.fire(HttpContext.java:282)
	at io.vertx.ext.web.client.impl.HttpContext.dispatchResponse(HttpContext.java:243)
	at io.vertx.ext.web.client.impl.HttpContext.lambda$null$2(HttpContext.java:373)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)

Error:  Tests run: 5, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 20.351 s <<< FAILURE! - in itest.BasicCommandChannelIT
Error:  itest.BasicCommandChannelIT.shouldGetScanTargetsResponse  Time elapsed: 20.033 s  <<< ERROR!
java.util.concurrent.TimeoutException
	at itest.BasicCommandChannelIT.shouldGetScanTargetsResponse(BasicCommandChannelIT.java:129)

[INFO] Running itest.ClientAssetsIT
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.264 s - in itest.ClientAssetsIT
[INFO] Running itest.GrafanaDatasourceIT
Warning:  Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.006 s - in itest.GrafanaDatasourceIT
[INFO] Running itest.RecordingWorkflowIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 9.772 s - in itest.RecordingWorkflowIT
[INFO] Running itest.ClientUrlIT
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.042 s - in itest.ClientUrlIT
[INFO] Running itest.NoopAuthIT
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.012 s - in itest.NoopAuthIT
[INFO] Running itest.UploadRecordingIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.985 s - in itest.UploadRecordingIT
[INFO] 
[INFO] Results:
[INFO] 
Error:  Errors: 
Error:    BasicCommandChannelIT.shouldGetScanTargetsResponse:129 » Timeout
[INFO] 
Error:  Tests run: 19, Failures: 0, Errors: 1, Skipped: 1
```

The test cleanup in question is here: https://github.com/cryostatio/cryostat/blob/v1/src/test/java/itest/ConnectToExternalTargetsIT.java#L77

It seems that this cleanup needs to wait until all of the test containers are observed to actually be stopped/destroyed before the method returns, otherwise a subsequent itest may run before the state cleanup is complete and unexpected containers may still be present in the pod.

--- a/pom.xml
+++ b/pom.xml
@@ -350,6 +350,7 @@
         </execution>
       </executions>
       <configuration>
+        <runOrder>random</runOrder>
         <systemPropertyVariables>
           <cryostatWebPort>${cryostat.itest.webPort}</cryostatWebPort>
           <cryostatPodName>${cryostat.itest.podName}</cryostatPodName>

--- a/src/test/java/itest/InterleavedExternalTargetRequestsIT.java
+++ b/src/test/java/itest/InterleavedExternalTargetRequestsIT.java
@@ -99,13 +99,19 @@ class InterleavedExternalTargetRequestsIT extends TestBase {
                 System.out.println("setup complete, continuing to tests");
                 break;
             } else if (numTargets < NUM_EXT_CONTAINERS + 1) {
-                System.err.println(String.format("%d targets found on attempt %d - waiting for setup to complete", numTargets, attempts + 1));
+                System.err.println(
+                        String.format(
+                                "%d targets found on attempt %d - waiting for setup to complete",
+                                numTargets, attempts + 1));
                 if (attempts > 6) {
                     throw new Exception("setup failed");
                 }
                 Thread.sleep(5_000);
             } else {
-                System.err.println(String.format("%d targets found on attempt %d - too many!", numTargets, attempts + 1));
+                System.err.println(
+                        String.format(
+                                "%d targets found on attempt %d - too many!",
+                                numTargets, attempts + 1));
                 throw new Exception("setup failed");
             }
             attempts++;
@@ -136,7 +142,10 @@ class InterleavedExternalTargetRequestsIT extends TestBase {
                             });
             int numTargets = resp.get(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS);
             if (numTargets > 1) {
-                System.err.println(String.format("%d targets found on attempt %d - waiting for teardown to complete", numTargets, attempts + 1));
+                System.err.println(
+                        String.format(
+                                "%d targets found on attempt %d - waiting for teardown to complete",
+                                numTargets, attempts + 1));
                 if (attempts++ > 6) {
                     throw new Exception("teardown failed");
                 }

--- a/src/test/java/itest/InterleavedExternalTargetRequestsIT.java
+++ b/src/test/java/itest/InterleavedExternalTargetRequestsIT.java
@@ -80,13 +80,73 @@ class InterleavedExternalTargetRequestsIT extends TestBase {
                                 .collect(Collectors.toList())
                                 .toArray(new CompletableFuture[0]))
                 .join();
-        Thread.sleep(7_500L); // wait for JDP to discover new containers
+
+        // Query every 5s for up to 6 queries, waiting until we have discovered the expected
+        // number of targets via JDP (NUM_EXT_CONTAINERS, + 1 for Cryostat itself).
+        int attempts = 0;
+        while (true) {
+            CompletableFuture<Integer> resp = new CompletableFuture<>();
+            webClient
+                    .get("/api/v1/targets")
+                    .send(
+                            ar -> {
+                                if (assertRequestStatus(ar, resp)) {
+                                    resp.complete(ar.result().bodyAsJsonArray().size());
+                                }
+                            });
+            int numTargets = resp.get(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+            if (numTargets == NUM_EXT_CONTAINERS + 1) {
+                System.out.println("setup complete, continuing to tests");
+                break;
+            } else if (numTargets < NUM_EXT_CONTAINERS + 1) {
+                System.err.println(String.format("%d targets found on attempt %d - waiting for setup to complete", numTargets, attempts + 1));
+                if (attempts > 6) {
+                    throw new Exception("setup failed");
+                }
+                Thread.sleep(5_000);
+            } else {
+                System.err.println(String.format("%d targets found on attempt %d - too many!", numTargets, attempts + 1));
+                throw new Exception("setup failed");
+            }
+            attempts++;
+        }
     }
 
     @AfterAll
     static void cleanup() throws Exception {
         for (String id : CONTAINERS) {
             Podman.kill(id);
+        }
+
+        // Query every 5s for up to 6 queries. If we still see additional targets other than
+        // the Cryostat instance itself, bail out - teardown failed. JDP discovery may take some
+        // time to notice that targets have disappeared after the processes/containers are killed,
+        // but this should be only a few seconds.
+        // https://github.com/cryostatio/cryostat/issues/501#issuecomment-856264316
+        int attempts = 0;
+        while (true) {
+            CompletableFuture<Integer> resp = new CompletableFuture<>();
+            webClient
+                    .get("/api/v1/targets")
+                    .send(
+                            ar -> {
+                                if (assertRequestStatus(ar, resp)) {
+                                    resp.complete(ar.result().bodyAsJsonArray().size());
+                                }
+                            });
+            int numTargets = resp.get(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+            if (numTargets > 1) {
+                System.err.println(String.format("%d targets found on attempt %d - waiting for teardown to complete", numTargets, attempts + 1));
+                if (attempts++ > 6) {
+                    throw new Exception("teardown failed");
+                }
+                Thread.sleep(5_000);
+            } else if (numTargets == 0) {
+                throw new Exception("teardown failed - all containers gone, including Cryostat");
+            } else {
+                System.out.println("teardown complete");
+                break;
+            }
         }
     }
 

